lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1672 @@
+		}
+		(self.nodes.len() as u64).write(writer)?;
+		for (ref node_id, ref node_info) in self.nodes.iter() {
+			node_id.write(writer)?;
+			node_info.write(writer)?;
+		}
+		Ok(())
+	}
+}
+
+impl Readable for NetworkGraph {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NetworkGraph, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = BTreeMap::new();
+		for _ in 0..channels_count {
+			let chan_id: u64 = Readable::read(reader)?;
+			let chan_info = Readable::read(reader)?;
+			channels.insert(chan_id, chan_info);
+		}
+		let nodes_count: u64 = Readable::read(reader)?;
+		let mut nodes = BTreeMap::new();
+		for _ in 0..nodes_count {
+			let node_id = Readable::read(reader)?;
+			let node_info = Readable::read(reader)?;
+			nodes.insert(node_id, node_info);
+		}
+		Ok(NetworkGraph {
+			channels,
+			nodes,
+		})
+	}
+}
+
+impl std::fmt::Display for NetworkGraph {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "Network map\n[Channels]\n")?;
+		for (key, val) in self.channels.iter() {
+			write!(f, " {}: {}\n", key, val)?;
+		}
+		write!(f, "[Nodes]\n")?;
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	/// Returns all known valid channels
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns all known nodes
+	pub fn get_nodes<'a>(&'a self) -> &'a BTreeMap<PublicKey, NodeInfo> { &self.nodes }
+
+	/// For an already known node (from channel announcements), update its stored properties from a given node announcement
+	/// Announcement signatures *are not checked* here and should be checked by a caller beforehand.
+	pub fn update_node_from_announcement(&mut self, msg: &msgs::NodeAnnouncement) -> Result<bool, LightningError> {
+		match self.nodes.get_mut(&msg.contents.node_id) {
+			None => Err(LightningError{err: "No existing channels for node_announcement", action: ErrorAction::IgnoreError}),
+			Some(node) => {
+				if let Some(node_info) = node.announcement_info.as_ref() {
+					if node_info.last_update  >= msg.contents.timestamp {
+						return Err(LightningError{err: "Update older than last processed update", action: ErrorAction::IgnoreError});
+					}
+				}
+
+				let should_relay = msg.contents.excess_data.is_empty() && msg.contents.excess_address_data.is_empty();
+				node.announcement_info = Some(NodeAnnouncementInfo {
+					features: msg.contents.features.clone(),
+					last_update: msg.contents.timestamp,
+					rgb: msg.contents.rgb,
+					alias: msg.contents.alias,
+					addresses: msg.contents.addresses.clone(),
+					announcement_message: if should_relay { Some(msg.clone()) } else { None },
+				});
+
+				Ok(should_relay)
+			}
+		}
+	}
+
+	/// For a new or already known (from previous announcement) channel, store or update channel info,
+	/// after making sure it corresponds to a real transaction on-chain.
```

Comment:
We dont do this? Thats what checked_utxo is for? checked_utxo's properties should be detailed, however.